rust/geoparquet/src/reader/geo_ext.rs

```diff
@@ -0,0 +1,184 @@
+    /// `true`. If you want to leave geometries as WKB, set it to `false`.
+    fn geoarrow_schema(
+        &self,
+        geo_metadata: &GeoParquetMetadata,
+        parse_to_native: bool,
+        coord_type: CoordType,
+    ) -> GeoArrowResult<SchemaRef>;
+
+    /// Construct a spatial [ArrowPredicate] that keeps rows that intersect with the provided
+    /// `bbox`.
+    ///
+    /// Note that this will **replace** any existing [`RowFilter`]s. If you want to use more than
+    /// one filter, you should create [`ArrowPredicate`]s directly and pass in your own
+    /// [`RowFilter`] to [`ArrowReaderBuilder::with_row_filter`].
+    ///
+    /// Note that the `bbox` must be in the same coordinate system as the geometries in the
+    /// designated geometry column.
+    ///
+    /// If `column_name` is `None`, the primary geometry column will be used.
+    fn intersecting_arrow_predicate(
+        &self,
+        bbox: Rect,
+        geo_metadata: &GeoParquetMetadata,
+        column_name: Option<&str>,
+    ) -> GeoArrowResult<Box<dyn ArrowPredicate>>;
+
+    /// Apply a spatial intersection [RowFilter] to this [`ArrowReaderBuilder`].
+    ///
+    /// Note that this will **replace** any existing [`RowFilter`]s on the builder. If you want to
+    /// use more than one [`ArrowPredicate`] in your [`RowFilter`], use
+    /// `Self::intersecting_arrow_predicate` to create the [`ArrowPredicate`] yourself. Then create
+    /// your own [`RowFilter`] that you pass to [`ArrowReaderBuilder::with_row_filter`].
+    ///
+    /// If `column_name` is `None`, the primary geometry column will be used.
+    fn with_intersecting_row_filter(
+        self,
+        bbox: Rect,
+        geo_metadata: &GeoParquetMetadata,
+        column_name: Option<&str>,
+    ) -> GeoArrowResult<Self>;
+
+    /// Find the row groups that intersect with the bounding box.
+    ///
+    /// Note that the `bbox` must be in the same coordinate system as the geometries in the
+    /// designated geometry column.
+    ///
+    /// If `column_name` is `None`, the primary geometry column will be used.
+    fn intersecting_row_groups(
+        &self,
+        bbox: Rect,
+        geo_metadata: &GeoParquetMetadata,
+        column_name: Option<&str>,
+    ) -> GeoArrowResult<Vec<usize>>;
+
+    /// Select row groups to read based on the bounding box.
+    ///
+    /// Note that this will **replace** any existing row group selection. If you want more detailed
+    /// selection of row groups, use [`ArrowReaderBuilder::with_row_groups`] yourself.
+    ///
+    /// If `column_name` is `None`, the primary geometry column will be used.
+    fn with_intersecting_row_groups(
+        self,
+        bbox: Rect,
+        geo_metadata: &GeoParquetMetadata,
+        column_name: Option<&str>,
+    ) -> GeoArrowResult<Self>;
+}
+
+impl<T> GeoParquetReaderBuilder for ArrowReaderBuilder<T> {
+    fn geoparquet_metadata(&self) -> Option<GeoParquetMetadata> {
```

Comment:
I'm thinking maybe this should return `Option<Result<GeoParquetMetadata>>`. Return `None` if there's no `geo` key; return `Some(Err)` if there is a `geo` key but it can't be parsed.

Reply:
You can use `transpose` easily to switch between `Result<Option<>>` and `Option<Result<>>`.

I think `Option<Result<>>` maps better semantically here, as _first_ you learn whether the `Option` is `Some` or `None` (because you can very easily check whether the `geo` Parquet key is set). And then only _second_ you learn whether the GeoParquet data is valid.